### PR TITLE
Module4 homework

### PR DESCRIPTION
**Pull request by alvarolealdz**
Tarea 4 lista, en estos tres dias me pongo con la 5 para haberme pegado bien con ella antes de la clase del domingo. Me he visto bien durante esta tarea aunque me gustaría haber podido echarle más horas. 

Conclusiones:

- Establecimiento de Benchmarks Claros: Iniciamos el proceso estableciendo puntos de referencia claros con resultados específicos a superar, lo que nos proporcionó objetivos definidos desde el principio.

- Importancia de Iterar sobre Parámetros Específicos: Me gustaría resaltar la importancia de iterar sobre los distintos parámetros específicos de cada modelo antes de comenzar a añadir o eliminar variables. De esta manera, podemos asegurarnos de que el modelo base esté optimizado y tenga el mejor rendimiento posible con la estructura de datos actual.

- Random Forest: Al construir nuestro modelo de Random Forestacercándonos a superar los benchmarks establecidos por los modelos lineales pero sin éxito. Este resultado podría atribuirse a la naturaleza intrínseca de los datos, a que solamente hemos trabajado con variables numéricas o las configuraciones específicas adoptadas para el modelo.

- Éxito con los Modelos de Gradient Boosting Trees (GBT):  Logramos una mejora significativa y superamos nuestros benchmarks mediante la implementación de modelos de Gradient Boosting Trees. Este éxito no solo demostró la eficacia de los modelos no lineales para nuestro conjunto de datos sino que también resultó en un modelo simplificado y más eficiente. Al reducir la complejidad y enfocarnos en las variables más significativas, pudimos mejorar la interpretabilidad sin sacrificar el rendimiento.

- Como aspectos a mejorar, me gustaría haber indagado más en el preprocesamiento de variables numéricas, como hicimos en la práctica anterior con StandardScaler, y haber incluido posibles variables categóricas que podrían contener información relevante.

### CODE REVIEW

File: src/module_4/module_4_model-fitting.md [hunk 1]
```diff
@@ -0,0 +1,1214 @@
...
+<p>5 rows × 27 columns</p>
+</div>
+
+
+
+
+```python
+min_products = 5
+order_size = df_feature_frame.groupby('order_id')['outcome'].sum()
+orders_of_min_size = order_size[order_size >= min_products].index
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:208]: I am not a big fan of using "index", it's not really useful and can be misleading. Instead, use your `order_id`. So in the line below do: `orders_of_min_size = order_size[order_size >= min_products]["order_id"]

File: src/module_4/module_4_model-fitting.md [hunk 2]
```diff
@@ -0,0 +1,1214 @@
...
+
+
+```python
+df_filtered = df_filtered.sort_values('created_at')
+
+order_ids = df_filtered['order_id'].unique()
+
+total_ids = len(order_ids)
+train_end = int(total_ids * 0.7)
+val_end = train_end + int(total_ids * 0.15)
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:231]: This evaluation schema is in the good direction but I would consider:

- make the train/eval/test split over days instead of over orders (bear in mind that in the train period you are NOT supposed to know how many orders you are gonna have in the future, while with the schema that you are proposing you are implicitly assuming this number, this is not true if you just consider number of days)
- Have a rolling window where you keep moving the train / test to the right and you train and evaluate in different months to see how the model behaves in different periods

File: src/module_4/module_4_model-fitting.md [hunk 3]
```diff
@@ -0,0 +1,1214 @@
...
+categorical_cols = ['product_type', 'vendor']
+binary_cols = ['ordered_before', 'abandoned_before', 'active_snoozed', 'set_as_regular']
+numerical_cols = [col for col in features_cols if col not in categorical_cols + binary_cols]
+```
+
+
+```python
+df_filtered = df_filtered.sort_values('created_at')
+
+order_ids = df_filtered['order_id'].unique()
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:227]: double check that the order of the `created_at`s is preserved after doing `unique`. In order to be super sure I would do: `df_filtered.sort_values('created_at').drop_duplicates("order_id"keep = "first")`

File: src/module_4/module_4_model-fitting.md [hunk 4]
```diff
@@ -0,0 +1,1214 @@
...
+train_end = int(total_ids * 0.7)
+val_end = train_end + int(total_ids * 0.15)
+
+train_order_ids = order_ids[:train_end]
+val_order_ids = order_ids[train_end:val_end]
+test_order_ids = order_ids[val_end:]
+
+train_df = df_filtered[df_filtered['order_id'].isin(train_order_ids)]
+val_df = df_feature_frame[df_feature_frame['order_id'].isin(val_order_ids)]
+test_df = df_feature_frame[df_feature_frame['order_id'].isin(test_order_ids)]
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:239]: put all this train / eval / test inside a function

File: src/module_4/module_4_model-fitting.md [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+    return df.drop(label_col, axis=1), df[label_col]
+
+X_train, y_train = feature_label_split(train_df, label_col)
+X_val, y_val = feature_label_split(val_df, label_col)
+X_test, y_test = feature_label_split(test_df, label_col)
+```
+
+
+```python
+"""""Función para crear los gráficos del performance de nuestro modelo"""
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:263]: check how to comment/document functions in Python: https://stackoverflow.com/questions/2357230/what-is-the-proper-way-to-comment-functions-in-python

File: src/module_4/module_4_model-fitting.md [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+    fi.sort_values(by="importance", ascending=True, inplace=True)
+    fi.plot(kind="barh", x="features", y="importance", ax=ax)
+    return fig, ax
+```
+
+## Previous Baselines
+
+### Initial Benchmark 
+
+Para nuestra evaluación inicial, utilizamos la característica de "popularidad global" para determinar si el pedido puede ser realizado o no.
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:316]: how is this global popularity computed? if it's taking into account orders of the WHOLE period then you are leaking data and shouldn't be used

File: src/module_4/module_4_model-fitting.md [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+X_test, y_test = feature_label_split(test_df, label_col)
+```
+
+
+```python
+"""""Función para crear los gráficos del performance de nuestro modelo"""
+def plot_metrics(model_name, y_pred, y_test, figsize=(10, 7)):
+    precision, recall, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall, precision)
+    fpr, tpr, _ = roc_curve(y_test, y_pred)
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:267]: You are using y_test instead of y_eval. The idea of y_test is that you only touch it at the END when you have chose the final model and you want to know the final performance

File: src/module_4/module_4_model-fitting.md [hunk 8]
```diff
@@ -0,0 +1,1214 @@
...
+for c in regularization_values:
+    lr = make_pipeline(
+        StandardScaler(),
+        LogisticRegression(penalty='l1',C=c, solver='liblinear')
+    )
+    lr.fit(X_train[train_cols], y_train)
+
+    y_proba = lr.predict_proba(X_test[train_cols])[:, 1]
+    
+    roc_auc = roc_auc_score(y_test, y_proba)
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:352]: same as before, you shouldn't use `y_test`, instead use `y_eval`

File: src/module_4/module_4_model-fitting.md [hunk 9]
```diff
@@ -0,0 +1,1214 @@
...
+    
+    if roc_auc > best_auc:
+        best_auc = roc_auc
+        best_c = c
+        best_model = lr
+
+y_pred = best_model.predict_proba(X_test[train_cols])[:, 1]
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:359]: instead of visualizing just the best_model metrics why don't you look at all of them? this way you can understand if there is a big difference / variance between the model results (and therefore finnetuning even further makes sense) or if doesn't really matter much

File: src/module_4/module_4_model-fitting.md [hunk 10]
```diff
@@ -0,0 +1,1214 @@
...
+}
+
+for n_trees in n_trees_grid:
+    rf = RandomForestClassifier(n_estimators=n_trees)
+    rf.fit(X_train[train_cols], y_train)
+    
+    train_probs = rf.predict_proba(X_train[train_cols])[:, 1]
+    val_probs = rf.predict_proba(X_val[train_cols])[:, 1]
+    
+    metrics['n_trees'].append(n_trees)
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:421]: you are doing something quite similar for every model, which in pseudocode can be summarized as: 
```
for hyperparam in hyperparam_set:
 model[hyperparam].train(X)
 pred_test = model[hyperparam].pred(X_test)
 metrics_train = metr_fnc(pred,y_test)
 pred_train = model[hyperparam].pred(X_train)
 metrics_train = metr_fnc(pred,y_train)
```
can you try to conceptualise this and put it on a function?

File: src/module_4/module_4_model-fitting.md [hunk 11]
```diff
@@ -0,0 +1,1214 @@
...
+for i in range(len(n_trees_grid)):
+    print(f"# Trees: {metrics['n_trees'][i]}")
+    print(f"Train AUC: {metrics['train_auc'][i]:.4f}, Validation AUC: {metrics['val_auc'][i]:.4f}")
+    print(f"Train AP: {metrics['train_ap'][i]:.4f}, Validation AP: {metrics['val_ap'][i]:.4f}")
+    print(f"Train Cross Entropy: {metrics['train_ce'][i]:.4f}, Validation Cross Entropy: {metrics['val_ce'][i]:.4f}")
+    print()
+```
+
+    Métricas:
+    # Trees: 5
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:439]: What if you plot these results into a plot to understand them better?

File: src/module_4/module_4_model-fitting.md [hunk 12]
```diff
@@ -0,0 +1,1214 @@
...
+
+
+Tras ejecutar este código, podemos observar los siguientes aspectos:
+
+- La elevada precisión obtenida en el conjunto de entrenamiento es un resultado esperado, considerando que estamos implementando un modelo de bosque aleatorio.
+- Se constata una mejora considerable en el rendimiento del modelo con el aumento en el número de árboles. Aunque sería factible continuar ajustando este parámetro, experimentando con valores como 150 o 200, las restricciones de tiempo nos llevan a establecer 100 como el número óptimo de árboles.
+
+### Model 1. Random Forest - Without categorical variables
+
+Iniciaremos el entrenamiento de un primer modelo utilizando los parámetros previamente seleccionados y enfocándonos únicamente en las variables numéricas. Por el momento, excluiremos las variables categóricas debido a su complejidad.
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:468]: What's the difference between this and the RF that you just fit before? you are choosing the best one? why?

File: src/module_4/module_4_model-fitting.md [hunk 13]
```diff
@@ -0,0 +1,1214 @@
...
+}
+</style><div id="sk-container-id-1" class="sk-top-container"><div class="sk-text-repr-fallback"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class="sk-container" hidden><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-1" type="checkbox" checked><label for="sk-estimator-id-1" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;&nbsp;RandomForestClassifier<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.4/modules/generated/sklearn.ensemble.RandomForestClassifier.html">?<span>Documentation for RandomForestClassifier</span></a><span class="sk-estimator-doc-link fitted">i<span>Fitted</span></span></label><div class="sk-toggleable__content fitted"><pre>RandomForestClassifier()</pre></div> </div></div></div></div>
+
+
+
+
+```python
+plot_feature_importance(rf, train_cols)
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:907]: As I told you in the lecture, using the native feature importance function of the random forest can be a quite bad idea

Comment (gsantam) [reply src/module_4/module_4_model-fitting.md:907]: Again, how is the `global_popularity` computed? Can it be a source of data leakage?

File: src/module_4/module_4_model-fitting.md [hunk 14]
```diff
@@ -0,0 +1,1214 @@
...
+n_trees_grid = [100] 
+
+for lr in learning_rates:
+    for depth in max_depths:
+        for n_trees in n_trees_grid:
+        
+            gbt = GradientBoostingClassifier(learning_rate=lr, max_depth=depth, n_estimators=n_trees)
+            gbt.fit(X_train[train_cols], y_train)
+
+            train_probs = gbt.predict_proba(X_train[train_cols])[:, 1]
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:1015]: Again, this looks like duplicated code with respect to other models

File: src/module_4/module_4_model-fitting.md [hunk 15]
```diff
@@ -0,0 +1,1214 @@
...
+
+Para continuar avanzando con la práctica, experimentaremos con modelos de Gradient Boosting Trees, utilizando los modelos anteriores como punto de referencia. Comenzaremos ajustando los distintos parámetros de nuestro modelo para identificar la configuración más óptima. Para simplificar el proceso, fijaremos el número de árboles en 100 como valor predeterminado para el modelo, ya que realizar iteraciones como en ocasiones anteriores consume demasiado tiempo.
+
+
+```python
+train_cols = numerical_cols + binary_cols
+
+learning_rates = [0.05, 0.1]
+max_depths = [1, 3, 5]
+n_trees_grid = [100]
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:1006]: Bear in mind that in order to train with 100 trees, you first need to train with 1, 2, 3... until 100. This means that you can know your performance for any number of trees up to 100. Can you use this to know what is the optimal number of trees that you need to use? Plot learning curves

File: src/module_4/module_4_model-fitting.md [hunk 16]
```diff
@@ -0,0 +1,1214 @@
...
+
+
+
+    (<Figure size 640x480 with 1 Axes>, <Axes: ylabel='features'>)
+
+
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_32_1.png)
```
Comment (gsantam) [inline src/module_4/module_4_model-fitting.md:1199]: I am missing a plot where you put all the metrics from all the different models to see how much do they change from one model to another one (is the best model just by chance or is clearly superior)

Comment (gsantam) [reply src/module_4/module_4_model-fitting.md:1199]: Also, at the end you should take the best model and evaluate it in test (which you are not supposed to touch until now)

**Review by gsantam — APPROVED** [review]

File: src/module_4/module_4_model-fitting.md (hunk 10 above)
Comment (alvarolealdz) [reply src/module_4/module_4_model-fitting.md:421]: hey Guille! Totally agree. A conceptualised function would look like:

Function:
def evaluate_model(model, X_train, y_train, X_test, y_test, metrics_function):
    model.fit(X_train, y_train)
    pred_test = model.predict(X_test)
    metrics_test = metrics_function(pred_test, y_test)
    pred_train = model.predict(X_train)
    metrics_train = metrics_function(pred_train, y_train)
    return metrics_test, metrics_train

Dictionary with the hyperparameters we want to test:
hyperparam_set = {...}
for hyperparam in hyperparam_set:
    model = MyModel(hyperparam)
    metrics_test, metrics_train = evaluate_model(
    model, X_train, y_train, X_test, y_test, some_metrics_function
    )
